plutus-core/testlib/PlutusIR/Properties/Typecheck.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+
+{- | Typechecking property tests for all PIR compiler passes. Currently we only typecheck the term
+ after a pass.
+
+ There were test failures when we infer the type before a pass and check that the
+ type after a pass is the same. The failures are probably caused by the
+ `PlutusIR.TypeCheck.AllowEscape` typechecker config.
+-}
+module PlutusIR.Properties.Typecheck where
+
+import Control.Monad.Reader
+import PlutusCore.Default
+import PlutusCore.Generators.QuickCheck.Utils
+import PlutusCore.Quote
+import PlutusCore.Rename
+import PlutusCore.TypeCheck qualified as PLC
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Provenance (original)
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusIR.TypeCheck
+import PlutusPrelude (($>))
+import Test.QuickCheck (Property, elements)
+import Test.QuickCheck.Arbitrary
+
+instance Arbitrary (BuiltinSemanticsVariant DefaultFun) where
+    arbitrary = elements [DefaultFunSemanticsVariant1, DefaultFunSemanticsVariant2]
+
+-- Convert Either Error () to Either String () to match with the Testable (Either String ())
+-- instance.
+convertToEitherString :: Either (Error DefaultUni DefaultFun ()) ()
+  -> Either String ()
+convertToEitherString = \case
+  Left err -> Left $ show err
+  Right () -> Right ()
+
+-- | Check that a term typechecks after one of the pure passes.
+pure_typecheck_prop ::
```

Comment:
we don't generally use snake-case?

Reply:
I was under the impression that we do in tests - maybe not.

Reply:
Take a look! We use it just for the prefix for tasty-discover, which I think is using snake-case for that part precisely because it's unusual.

Reply:
I don't think it's just `tasty-discover`, if you look at the `QuickCheck` documentation it uses snake-case too. I'm happy to stick to only using snake-case for tests/properties for `tasty-discover` though.